libretro/libretro.cpp

```diff
@@ -411,58 +424,78 @@ void retro_set_input_state(retro_input_state_t cb) { input_state_cb = cb; }
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_RIGHT, "D-Pad Right" },
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_B, "Cross" },
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_A, "Circle" },
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_X, "Triangle" },
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_Y, "Square" },
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_L, "L" },
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_R, "R" },
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_SELECT, "Select" },
+      { 0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_START, "Start" },
+      { 0, RETRO_DEVICE_ANALOG, RETRO_DEVICE_INDEX_ANALOG_RIGHT, RETRO_DEVICE_ID_ANALOG_X, "Right Analog X" },
+      { 0, RETRO_DEVICE_ANALOG, RETRO_DEVICE_INDEX_ANALOG_RIGHT, RETRO_DEVICE_ID_ANALOG_Y, "Right Analog Y" },
+      { 0, RETRO_DEVICE_ANALOG, RETRO_DEVICE_INDEX_ANALOG_LEFT, RETRO_DEVICE_ID_ANALOG_X, "Left Analog X" },
+      { 0, RETRO_DEVICE_ANALOG, RETRO_DEVICE_INDEX_ANALOG_LEFT, RETRO_DEVICE_ID_ANALOG_Y, "Left Analog Y" },
+      { 0 },
+   };
+   environ_cb(RETRO_ENVIRONMENT_SET_INPUT_DESCRIPTORS, desc);
+
+   if (environ_cb(RETRO_ENVIRONMENT_GET_INPUT_BITMASKS, NULL))
+      libretro_supports_bitmasks = true;
 
+   struct retro_log_callback log;
    if (environ_cb(RETRO_ENVIRONMENT_GET_LOG_INTERFACE, &log))
    {
+      LogManager::Init(&g_Config.bEnableLogging);
       printfLogger = new PrintfLogger(log);
-      LogManager *logman = LogManager::GetInstance();
+      LogManager* logman = LogManager::GetInstance();
       logman->RemoveListener(logman->GetConsoleListener());
       logman->RemoveListener(logman->GetDebuggerListener());
       logman->ChangeFileLog(nullptr);
       logman->AddListener(printfLogger);
-#if 1
       logman->SetAllLogLevels(LogTypes::LINFO);
-#endif
    }
 
-   if (environ_cb(RETRO_ENVIRONMENT_GET_INPUT_BITMASKS, NULL))
-      libretro_supports_bitmasks = true;
+   g_Config.Load("", "");
+   g_Config.iInternalResolution = 0;
+   g_Config.sMACAddress = "12:34:56:78:9A:BC";
+
+   const char* nickname = NULL;
+   if (environ_cb(RETRO_ENVIRONMENT_GET_USERNAME, &nickname) && nickname)
+      g_Config.sNickName = std::string(nickname);
+
+   Path retro_base_dir;
+   Path retro_save_dir;
+   const char* dir_ptr = NULL;
+   if (environ_cb(RETRO_ENVIRONMENT_GET_SYSTEM_DIRECTORY, &dir_ptr) && dir_ptr)
+      retro_base_dir = Path(dir_ptr);
+
+   if (environ_cb(RETRO_ENVIRONMENT_GET_SAVE_DIRECTORY, &dir_ptr) && dir_ptr)
+      retro_save_dir = Path(dir_ptr);
+
+   retro_base_dir /= "PPSSPP";
+
+   g_Config.currentDirectory = retro_base_dir;
+   g_Config.defaultCurrentDirectory = retro_base_dir;
+   g_Config.memStickDirectory = retro_save_dir;
+   g_Config.flash0Directory = retro_base_dir / "flash0";
+   g_Config.internalDataDirectory = retro_base_dir;
+
+   VFSRegister("", new DirectoryAssetReader(retro_base_dir));
+
+   host = new LibretroHost;
```

Comment:
Code style and safety thing - when using the new operator, always include the parenthesis.

`host = new LibretroHost();`

Reply:
And done.